wm/frame.go

```diff
@@ -3,271 +3,214 @@
-	for _, w := range s.frames {
-		if w.(*frame).id == id || w.(*frame).win == id {
-			return w
-		}
-	}
-
-	return nil
-}
-
-func (f *frame) unFrame() {
-	if !f.framed {
-		return
-	}
-
-	frame := f.wm.frameForWin(f.win)
-	f.wm.RemoveWindow(frame)
-
-	if frame == nil {
+func (fr *frame) unFrame() {
+	if !fr.framed {
 		return
 	}
 
-	xproto.ReparentWindow(f.wm.x.Conn(), f.win, f.wm.x.RootWin(), f.x, f.y)
-	xproto.UnmapWindow(f.wm.x.Conn(), f.id)
-}
-
-func (f *frame) Decorated() bool {
-	return f.framed
-}
-
-func (f *frame) Title() string {
-	if f.title == "" {
-		f.title = windowName(f.wm.x, f.win)
-	}
-
-	return f.title
-}
-
-func (f *frame) Class() []string {
-	return windowClass(f.wm.x, f.win)
-}
-
-func (f *frame) Command() string {
-	return windowCommand(f.wm.x, f.win)
-}
-
-func (f *frame) IconName() string {
-	return windowIconName(f.wm.x, f.win)
-}
-
-func (f *frame) Close() {
-	winProtos, err := icccm.WmProtocolsGet(f.wm.x, f.win)
-	if err != nil {
-		log.Println("GetProtocols err", err)
-	}
-
-	askNicely := false
-	for _, proto := range winProtos {
-		if proto == "WM_DELETE_WINDOW" {
-			askNicely = true
-		}
-	}
-
-	if !askNicely {
-		err := xproto.DestroyWindowChecked(f.wm.x.Conn(), f.win).Check()
-		if err != nil {
-			log.Println("Close Err", err)
-		}
-
-		return
-	}
+	cli := fr.cli.wm.clientForWin(fr.cli.win)
+	fr.cli.wm.RemoveWindow(cli)
 
-	protocols, err := xprop.Atm(f.wm.x, "WM_PROTOCOLS")
-	if err != nil {
+	if cli == nil {
 		return
 	}
 
-	delWin, err := xprop.Atm(f.wm.x, "WM_DELETE_WINDOW")
-	if err != nil {
-		return
-	}
-	cm, err := xevent.NewClientMessage(32, f.win, protocols,
-		int(delWin))
-	err = xproto.SendEventChecked(f.wm.x.Conn(), false, f.win, 0,
-		string(cm.Bytes())).Check()
-	if err != nil {
-		log.Println("WinDelete Err", err)
+	if fr != nil {
+		xproto.ReparentWindow(fr.cli.wm.x.Conn(), fr.cli.win, fr.cli.wm.x.RootWin(), fr.x, fr.y)
 	}
+	xproto.UnmapWindow(fr.cli.wm.x.Conn(), fr.cli.id)
 }
 
-func (f *frame) Focus() {
-	xproto.SetInputFocus(f.wm.x.Conn(), 0, f.win, 0)
-}
-
-func (f *frame) press(x, y int16) {
-	f.mouseX = x
-	f.mouseY = y
-	if x >= int16(f.width-f.wm.buttonWidth()) && y >= int16(f.height-f.wm.buttonWidth()) {
-		f.mouseResize = true
+func (fr *frame) press(x, y int16) {
+	fr.mouseX = x
+	fr.mouseY = y
+	if x >= int16(fr.width-fr.cli.wm.buttonWidth()) && y >= int16(fr.height-fr.cli.wm.buttonWidth()) {
+		fr.mouseResize = true
 	} else {
-		f.mouseResize = false
+		fr.mouseResize = false
 	}
 
-	f.wm.RaiseToTop(f)
+	fr.cli.wm.RaiseToTop(fr.cli)
 }
 
-func (f *frame) release(x, y int16) {
-	if x > int16(f.wm.borderWidth()) && x < int16(f.wm.borderWidth()+f.wm.buttonWidth()) &&
-		y < int16(f.wm.borderWidth()+f.wm.titleHeight()) {
-		f.Close()
+func (fr *frame) release(x, y int16) {
+	if x > int16(fr.cli.wm.borderWidth()) && x < int16(fr.cli.wm.borderWidth()+fr.cli.wm.buttonWidth()) &&
+		y < int16(fr.cli.wm.borderWidth()+fr.cli.wm.titleHeight()) {
```

Comment:
The y coordinate is the same for all of these - check it at the top and this gets less complex